Combine interpolating fields in one operation

gassmoeller opened the pull request:
@Wang-yijun and @KiralyAgi showed me a model with 36 prescribed compositional fields (a lot of components of a larger tensor), in which the interpolation from material model output into compositional fields took >50% of the total model runtime. This made me look into the algorithm and search for optimization potential. When we prescribe compositional fields from the material model output we currently do the following steps consecutively for each compositional field:

- evaluate the solution at the support points of the current compositional field
- evaluate the material model at the support points incl. prescribed outputs for all fields
- only write the material model output for the current field into the solution vector

All steps of this operation could be combined for all compositional fields that need to be interpolated, which would save significant time for models with many prescribed fields and expensive material models. This PR implements this change by copying the approach we already take for particles. We essentially collect all fields that are prescribed while looping through the compositional fields and then interpolate all of them together after all other fields have been solved.

Some intricacies:
- the order in which we solve/interpolate compositional fields now changes if prescribed fields are mixed in between "normal" fields, but this doesnt matter much, because the current linearization point for all fields is only updated once all fields are solved (to avoid inconsistent time stepping / reactions between them). So the reordering of fields will only affect the screen and statistics output
- the fields may have different finite elements. I solved this the same way we do in the function `compute_reactions`, by collecting all support points of all fields and creating a combined quadrature for all of them. I refactored the code from `compute_reactions` into its own function to avoid duplicating it.
- if we interpolate several compositional fields at the same time we need to change the screen output, it now no longer prints the field name

First tests by @Wang-yijun (and I could reproduce this) showed that this change reduces the time spent in this algorithm from >50% to <5% of the total model time for her model.

bobmyhill commented on source/simulator/helper_functions.cc:
```diff
@@ -2109,89 +2151,85 @@ namespace aspect
-    if (adv_field.is_temperature() ||
-        adv_field.compositional_variable != introspection.find_composition_type(CompositionalFieldDescription::density))
-      current_constraints.distribute (distributed_vector);
+        if (adv_field.is_temperature() ||
```
Can you add a comment to explain the logic behind this if statement?

bobmyhill reviewed (commented):
Nice! Just a couple of comments.

tjhei commented:
Thank you for cleaning up the code I wrote. 😉

gassmoeller commented:
Thanks for the reviews! I have addressed the comments as far as they are connected to this PR. I will open a separate PR to adress the two issues with global statistics.

> I know this existed in the code prior to this PR, but why is the code returning the max 32-bit unsigned integer for "composition solver 2"?

This is caused by behavior in the global_statistics postprocessor. If we do not solve an advection field (e.g. because we prescribe it) it reports the number of iterations as `numbers::invalid_unsigned_int` which is `std::numeric_limits<unsigned int>::max() - 1` (=4294967295). I am also not sure why the one test has a slightly different number (4294967291). I will address this in a separate PR.

> What is the reason for this change in ordering?

This is also caused by behavior in the global_statistics postprocessor. It reports the statistics columns in the order the advection fields have been solved. Because we now group all prescribed fields and interpolate them in a combined operation after all other fields this order has changed. I am actually not convinced that reporting the columns in the order of solving the fields makes much sense, we should just report temperature first, then all other fields in order. So far the two options were the same, but now they arent any more.

If you are ok with me addressing the issue with global_statistics in a separate PR this should be good to go.
